File: ggo-0.2/src/ggo/parser/SGFWriter.java
/*
 *  SGFWriter.java
 *
 *  gGo
 *  Copyright (C) 2002  Peter Strempel <pstrempel@t-online.de>
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */
package ggo.parser;

import ggo.*;

/**
 *  A subclass of Writer. This class converts a game into SGF format and saves the
 *  SGF string as the specified file.
 *
 *@author     Peter Strempel
 *@version    $Revision: 1.1 $, $Date: 2002/10/05 11:22:49 $
 */
public class SGFWriter extends Writer {
    private StringBuffer stream;
    private boolean isRoot;

    //{{{ doWrite() method
    /**
     *  Writes a given game tree to a string in sgf format.
     *
     *@return    String in sgf format
     *@see       ggo.parser.Writer#doWrite()
     */
    String doWrite() {
        Move root;
        try {
            root = boardHandler.getTree().getRoot();
        } catch (NullPointerException e) {
            System.err.println("SGFWriter.doWrite() - No game to save: " + e);
            e.printStackTrace();
            return null;
        }

        GameData data = boardHandler.getGameData();
        if (data == null) {
            System.err.println("SGFParser.doWrite() - No game data!");
            return null;
        }

        stream = new StringBuffer();

        // Traverse the tree recursive in pre-order
        isRoot = true;
        traverse(root, data);
        stream.append("\n");

        return stream.toString();
    } //}}}

    //{{{ writeGameHeader() method
    /**
     *  Print the sgf headers to the stream
     *
     *@param  data  Pointer of the GameData object
     */
    private void writeGameHeader(GameData data) {
        try {
            // Assemble data for root node
            stream.append(";GM[1]FF[4]" +  // We play Go, we use FF 4
            "CA[" + data.charset + "]" +  // Charset
            "AP[" + PACKAGE + ":" + VERSION + "]" // Application name : Version
             + "ST[1]"); // We show vars of current node
            if (data.gameName != null && data.gameName.length() > 0) // Skip game name if empty
                stream.append("GN[" + data.gameName + "]\n"); // Game Name
            else
                stream.append("\n");

            stream.append("SZ[" + data.size + "]" +  // Board size
            (data.handicap > 0 ? "HA[" + data.handicap + "]" : "") +  // Handicap
            "KM[" + data.komi + "]\n"); // Komi

            if (data.playerWhite != null && data.playerWhite.length() > 0)
                stream.append("PW[" + data.playerWhite + "]"); // White name

            if (data.rankWhite != null && data.rankWhite.length() > 0)
                stream.append("WR[" + data.rankWhite + "]"); // White rank

            if (data.playerBlack != null && data.playerBlack.length() > 0)
                stream.append("PB[" + data.playerBlack + "]"); // Black name

            if (data.rankBlack != null && data.rankBlack.length() > 0)
                stream.append("BR[" + data.rankBlack + "]"); // Black rank

            if (data.result != null && data.result.length() > 0)
                stream.append("RE[" + data.result + "]"); // Result

            if (data.date != null && data.date.length() > 0)
                stream.append("DT[" + data.date + "]"); // Date

            if (data.place != null && data.place.length() > 0)
                stream.append("PC[" + data.place + "]"); // Place

            if (data.copyright != null && data.copyright.length() > 0)
                stream.append("CP[" + data.copyright + "]"); // Copyright

            stream.append("\n");
        } catch (NullPointerException e) {
            System.err.println("Failed to write game: " + e);
        }
    } //}}}

    //{{{ traverse() method
    /**
     *  Method to traverse the tree and write the moves to sgf
     *
     *@param  t     Current move
     *@param  data  Pointer to the GameData object
     */
    private void traverse(Move t, GameData data) {
        stream.append("(");
        do {
            if (isRoot) {
                writeGameHeader(data);
                stream.append(t.saveMove(isRoot));
                isRoot = false;
            }
            else
                stream.append(t.saveMove(false));

            Move tmp = t.son;
            if (tmp != null && tmp.brother != null) {
                do {
                    stream.append("\n");
                    traverse(tmp, null);
                } while ((tmp = tmp.brother) != null);
                break;
            }
        } while ((t = t.son) != null);
        stream.append("\n)");
    } //}}}
}

File: glGo-0.0.6/src/sgf/sgfwriter.cpp
/*
 * sgfwriter.cpp
 *
 * $Id: sgfwriter.cpp,v 1.9 2003/10/05 02:12:15 peter Exp $
 *
 * glGo, a prototype for a 3D Goban based on wxWindows, OpenGL and SDL.
 * Copyright (c) 2003, Peter Strempel <pstrempel@gmx.de>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */

#ifdef __GNUG__
#pragma implementation "sgfwriter.h"
#endif

// For compilers that support precompilation, includes "wx.h".
#include "wx/wxprec.h"

#ifdef __BORLANDC__
#pragma hdrstop
#endif

#ifndef WX_PRECOMP
#include <wx/log.h>
#include <wx/intl.h>
#include <wx/msgdlg.h>
#endif

#include <wx/wfstream.h>
#include "sgfwriter.h"
#include "boardhandler.h"
#include "game.h"
#include "move.h"
#include "gamedata.h"


SGFWriter::SGFWriter()
{
    ostream = NULL;
}

bool SGFWriter::saveSGF(const wxString &filename, Game *game)
{
    Move *root = game->getRoot();
    wxASSERT(root != NULL);

    // This should actually never happen, but who knows.
    if (root == NULL)
    {
        wxMessageBox(_("Error saving SGF: The game has no root move!"),
                     _("SGF Error"), wxOK | wxICON_ERROR);
        return false;
    }

    // Create output file
    wxFileOutputStream fostream(filename);
    if (!fostream.Ok())
        return false;

    // Created buffered stream
    ostream = new wxBufferedOutputStream(fostream);
    wxASSERT(ostream != NULL);

    // Write game header to root node
    writeGameHeader(game->getGameData());

    // Traverse the tree recursive in pre-order
    traverse(root);
    ostream->Write("\n", 1);

    // Delete stream (this syncs the stream)
    delete ostream;
    ostream = NULL;

    return true;
}

void SGFWriter::writeGameHeader(GameData *data)
{
    wxString s;

    // We play Go, we use FF 4
    s = "(;GM[1]FF[4]";

    // Charset TODO
    // s += "CA[" + data.charset + "]" +

    // Application name : Version
    s += wxString::Format("AP[%s:%s]", PACKAGE , VERSION);

    // We show vars of current node
    s += "ST[1]\n";

    // Game Name
    if (!data->gameName.empty())
        s += _T("GN[") + data->gameName + "]";

    // Board size
    s += wxString::Format("\nSZ[%u]", data->board_size);

    // Handicap
    if (data->handicap >= 2)
        s += wxString::Format("HA[%u]", data->handicap);

    // Komi
    s += wxString::Format("KM[%.1f]\n", data->komi);

    // White name
    if (!data->whiteName.empty())
        s += _T("PW[") + data->whiteName + "]";

    // White rank
    if (!data->whiteRank.empty())
        s += _T("WR[") + data->whiteRank + "]";

    // Black name
    if (!data->blackName.empty())
        s += _T("PB[") + data->blackName + "]";

    // Black rank
    if (!data->blackRank.empty())
        s += _T("BR[") + data->blackRank + "]";

    // Result
    if (!data->result.empty())
        s += _T("RE[") + data->result + "]";

    // Time
    if (!data->time.empty())
        s += _T("TM[") + data->time + "]";

    // Date
    if (!data->date.empty())
        s += _T("DT[") + data->date + "]";

    // Place
    if (!data->place.empty())
        s += _T("PC[") + data->place + "]";

    // Copyright
    if (!data->copyright.empty())
        s += _T("CP[") + data->copyright + "]";

    s += "\n";

    ostream->Write(s.c_str(), s.Length());
}

void SGFWriter::traverse(Move *m)
{
    wxString s;
    if (m->parent != NULL)
        ostream->Write("\n(", 2);
    while (true)
    {
        s = m->MoveToSGF();
        ostream->Write(s.c_str(), s.Length());

        Move *tmp = m->son;
        if (tmp != NULL && tmp->brother != NULL)
        {
            do
            {
                traverse(tmp);
            } while ((tmp = tmp->brother) != NULL);
            break;
        }
        if ((m = m->son) != NULL)
            ostream->Write("\n", 1);
        else
            break;
    }
    ostream->Write(")", 1);
}
